### PR TITLE
Document HTTPX kwargs

### PR DESCRIPTION
**Pull request by Kludex**

### CODE REVIEW

File: logfire/_internal/integrations/httpx.py [hunk 1]
```diff
@@ -14,15 +20,9 @@
...
-    from typing import Awaitable, Callable, TypedDict, Unpack
+    from typing import TypedDict, Unpack
 
     import httpx
-    from opentelemetry.trace import Span
-
-    RequestHook = Callable[[Span, httpx.Request], None]
-    ResponseHook = Callable[[Span, httpx.Request, httpx.Response], None]
-    AsyncRequestHook = Callable[[Span, httpx.Request], Awaitable[None]]
-    AsyncResponseHook = Callable[[Span, httpx.Request, httpx.Response], Awaitable[None]]
```
Comment (Kludex) [inline logfire/_internal/integrations/httpx.py:25]: It's not `httpx.Request`. It's a `RequestInfo` which is a NamedTuple (???)...

**Comment by alexmojaki** [comment]
From https://github.com/pydantic/logfire/issues/655#issue-2727407545:

> Also [HTTPXInstrumentKwargs](https://github.com/pydantic/logfire/blob/743fb5d720ffa8b6b99d3f53a56d54bf10c5d8d5/logfire/_internal/integrations/httpx.py#L27-L32) is not included in docs which makes it basically undocumented.

I think the desire was for `HTTPXInstrumentKwargs` to actually link somewhere like other type hints. I know that this would be tricky, but as a compromise I think the docstring in the `Logfire` class needs more info, at least a link to the main integration page.

File: docs/integrations/httpx.md [hunk 2]
```diff
@@ -63,6 +63,152 @@ Let's see a minimal example below. You can run it with `python main.py`:
     asyncio.run(main())
     ```
 
+## Hooks
+
+The `logfire.instrument_httpx()` also accepts the following hooks:
```
Comment (alexmojaki) [inline docs/integrations/httpx.md]:
```suggestion
The `logfire.instrument_httpx()` method also accepts the following hooks:
```

File: docs/integrations/httpx.md [hunk 3]
```diff
@@ -63,6 +63,152 @@ Let's see a minimal example below. You can run it with `python main.py`:
...
+
+import logfire
+
+
+def set_request_headers(span: Span, request: RequestInfo):
+    method, url, headers, stream, extensions = request
+    attributes: defaultdict[str, str] = defaultdict(list)
+    for key, value in headers.items():
+        key = key.lower()
+        attributes[f'http.request.header.{key}'].append(value)
```
Comment (alexmojaki) [inline docs/integrations/httpx.md:97]: I started working on capturing headers. Currently that code is stashed. Once pushed this will become a bad example.

**Comment by Kludex** [comment]
> From [#655 (comment)](https://github.com/pydantic/logfire/issues/655#issue-2727407545):
> 
> > Also [HTTPXInstrumentKwargs](https://github.com/pydantic/logfire/blob/743fb5d720ffa8b6b99d3f53a56d54bf10c5d8d5/logfire/_internal/integrations/httpx.py#L27-L32) is not included in docs which makes it basically undocumented.
> 
> I think the desire was for `HTTPXInstrumentKwargs` to actually link somewhere like other type hints. I know that this would be tricky, but as a compromise I think the docstring in the `Logfire` class needs more info, at least a link to the main integration page.

So... You mean to create a `logfire.integrations.httpx` module? Not internal? I'm not sure I understand... 🤔

File: docs/integrations/httpx.md [hunk 4]
```diff
@@ -63,6 +63,152 @@ Let's see a minimal example below. You can run it with `python main.py`:
...
+    span.set_attributes(attributes)
+
+
+def request_hook(span: Span, request: RequestInfo):
+    """This is called when using a `httpx.Client`."""
+    set_request_headers(span, request)
+
+
+async def async_request_hook(span: Span, request: RequestInfo):
+    """This is called when using a `httpx.AsyncClient`."""
```
Comment (alexmojaki) [inline docs/integrations/httpx.md:107]: Two additional notes:

1. `async_request_hook` has to be a coroutine function, i.e. defined with `async def`, even if it only does sync things.
2. When instrumenting a client, the parameter is always called `request_hook`. To make it an async hook, it has to be defined as async. The type of client isn't taken into account.

I think we can automate dealing with this in the SDK. Maybe it should be an OTEL issue.

**Comment by alexmojaki** [comment]
I think the original request was that `HTTPXInstrumentKwargs` itself should

> > From [#655 (comment)](https://github.com/pydantic/logfire/issues/655#issue-2727407545):
> > > Also [HTTPXInstrumentKwargs](https://github.com/pydantic/logfire/blob/743fb5d720ffa8b6b99d3f53a56d54bf10c5d8d5/logfire/_internal/integrations/httpx.py#L27-L32) is not included in docs which makes it basically undocumented.
> > 
> > 
> > I think the desire was for `HTTPXInstrumentKwargs` to actually link somewhere like other type hints. I know that this would be tricky, but as a compromise I think the docstring in the `Logfire` class needs more info, at least a link to the main integration page.
> 
> So... Should we create the `logfire.integrations.httpx` module? Not internal... 🤔

That's one possibility. But I don't know what you'd do about types like `RequestHook`. We could type it ourselves instead of importing from OTEL. They have https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/httpx/httpx.html#opentelemetry.instrumentation.httpx.RequestInfo but not `RequestHook` as a type in docs. Also their `RequestInfo` is actually wrong, specifically `headers` is actually `httpx.Headers`.

But we don't *need* to have formal docs of all the types. I'm just saying that the docstring needs *something*. And no, it doesn't need to block this PR.

File: docs/integrations/httpx.md (hunk 3 above)
Comment (Kludex) [reply docs/integrations/httpx.md:97]: Can you suggest a useful example? I couldn't think of any...

File: docs/integrations/httpx.md [hunk 5]
```diff
@@ -63,6 +63,152 @@ Let's see a minimal example below. You can run it with `python main.py`:
...
+- `async_request_hook`: called before sending an async request.
+- `async_response_hook`: called after receiving an async response.
+
+You can use these hooks to add custom attributes that are not automatically captured by **Logfire**.
+
+### (Async) Request Hook
+
+The `request_hook` and `async_request_hook` hooks are called before sending a request. They receive two arguments:
+
+```py title="main.py" hl_lines="24-27"
```
Comment (alexmojaki) [inline docs/integrations/httpx.md:81]:
```suggestion
```py title="main.py"
```

File: docs/integrations/httpx.md [hunk 6]
```diff
@@ -63,6 +63,152 @@ Let's see a minimal example below. You can run it with `python main.py`:
...
+
+    Once a generator is consumed, it cannot be consumed again. This means that if you consume the stream
+    in the `request_hook` or `async_request_hook`, the request will fail.
+
+
+### (Async) Response Hook
+
+The `response_hook` and `async_response_hook` hooks are called after receiving a response. They receive three arguments:
+
+```py title="main.py" hl_lines="29-32"
```
Comment (alexmojaki) [inline docs/integrations/httpx.md:141]:
```suggestion
```py title="main.py"
```

File: docs/integrations/httpx.md [hunk 7]
```diff
@@ -63,6 +63,152 @@ Let's see a minimal example below. You can run it with `python main.py`:
...
+asyncio.run(main())
+```
+
+`RequestInfo` is a [`NamedTuple`][typing.NamedTuple] with the following fields:
+
+- `method`: The HTTP method of the request.
+- `url`: The URL of the request.
+- `headers`: The headers of the request.
+- `stream`: The stream of the request.
+- `extensions`: The extensions of the request.
```
Comment (alexmojaki) [inline docs/integrations/httpx.md]:
```suggestion
`RequestInfo` is a [`NamedTuple`][typing.NamedTuple] with the following attributes: `method`, `url`, `headers`, `stream`, and `extensions`.
```

Maybe I'm just fussy, but if we don't have more useful descriptions of these fields then I don't like the redundancy.

File: docs/integrations/httpx.md (hunk 5 above)
Comment (Kludex) [reply docs/integrations/httpx.md:81]: Why?

File: docs/integrations/httpx.md [hunk 8]
```diff
@@ -63,6 +63,146 @@ Let's see a minimal example below. You can run it with `python main.py`:
...
+
+asyncio.run(main())
+```
+
+`ResponseInfo` is a [`NamedTuple`][typing.NamedTuple] with the following fields:
+
+- `status_code`: The status code of the response.
+- `headers`: The headers of the response.
+- `stream`: The stream of the response.
+- `extensions`: The extensions of the response.
```
Comment (Kludex) [inline docs/integrations/httpx.md]:
```suggestion
`ResponseInfo` is a [`NamedTuple`][typing.NamedTuple] with the following attributes: `status_code`, `headers`, `stream`, and `extensions`.
```

File: docs/integrations/httpx.md (hunk 5 above)
Comment (alexmojaki) [reply docs/integrations/httpx.md:81]: Why are those lines highlighted? It looks weird. I thought it wasn't intentional.

File: docs/integrations/httpx.md (hunk 3 above)
Comment (alexmojaki) [reply docs/integrations/httpx.md:97]: Capturing URL query params as proper attributes?

**Comment by Kludex** [comment]
This PR is outdated. Let me craft something better..

**Pull request closed without merging**